File: solutions/c/dnd-character/1/dnd_character.c
#include "dnd_character.h"

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int ability(void)
{
    int min = 3;
    int max = 18;

    srand((unsigned)time(NULL));

    return rand() % (max - min + 1) + min;
}

int modifier(int score)
{
    return (score - 10) % 2 == -1 ? (score - 10) / 2 - 1 : (score - 10) / 2;
}

dnd_character_t make_dnd_character(void)
{
    dnd_character_t result;
    result.strength = ability();
    result.dexterity = ability();
    result.constitution = ability();
    result.intelligence = ability();
    result.wisdom = ability();
    result.charisma = ability();
    result.hitpoints = 10 + modifier(result.constitution);

    return result;
}

File: solutions/java/dnd-character/1/src/main/java/DnDCharacter.java
import java.util.ArrayList;
import java.util.List;

final class DnDCharacter {

    private final int strength = ability(rollDice());
    private final int dexterity = ability(rollDice());
    private final int constitution = ability(rollDice());
    private final int intelligence = ability(rollDice());
    private final int wisdom = ability(rollDice());
    private final int charisma = ability(rollDice());

    int ability(final List<Integer> scores) {
        int min = Integer.MAX_VALUE;
        int sum = 0;
        for (int score : scores) {
            if (score < min) {
                min = score;
            }
            sum += score;
        }

        return sum - min;
    }

    List<Integer> rollDice() {
        var results = new ArrayList<Integer>();
        for (int i = 0; i < 4; i++) {
            results.add((int) (Math.random() * 6) + 1);
        }

        return results;
    }

    int modifier(int input) {
        return Math.floorDiv(input - 10, 2);
    }

    int getStrength() {
        return strength;
    }

    int getDexterity() {
        return dexterity;
    }

    int getConstitution() {
        return constitution;
    }

    int getIntelligence() {
        return intelligence;
    }

    int getWisdom() {
        return wisdom;
    }

    int getCharisma() {
        return charisma;
    }

    int getHitpoints() {
        return 10 + modifier(constitution);
    }
}
